src/deploy/functions/cache/hash.ts

```diff
@@ -0,0 +1,55 @@
+import * as fs from "fs";
+import * as crypto from "crypto";
+import * as secrets from "../../../functions/secrets";
+import { Backend, Endpoint, EnvironmentVariables, allEndpoints } from "../backend";
+
+/**
+ * Retrieves the unique hash given a {@link Backend} and sourceFile.
+ * @param backend Backend of a set of functions
+ * @param sourceFile Packaged file contents of functions
+ */
+export async function getBackendHash(backend: Backend, sourceFile?: string) {
+  const hash = crypto.createHash("sha256");
+
+  // If present, hash the contents of the source file
+  if (sourceFile) {
+    await new Promise((resolve, reject) => {
+      const readStream = fs.createReadStream(sourceFile);
+      readStream.on("error", (err) => reject(err));
+      readStream.on("data", (chunk) => chunk && hash.push(chunk));
+      readStream.on("end", () => resolve(hash.digest("hex")));
+    });
+  }
+
+  // Hash the contents of the dotenv variables
+  if (hasEnvironmentVariables(backend.environmentVariables)) {
+    hash.push(JSON.stringify(backend.environmentVariables));
+  }
+
+  // Hash the secret versions.
+  const secretVersions = getSecretVersions(backend);
+  if (hasSecretVersions(secretVersions)) {
+    hash.push(JSON.stringify(secretVersions));
+  }
+
+  return hash.read().toString("hex");
```

Comment:
q: When I read documentation for crypto module, I expected `hash.digest('hex')`.  Is this roughly the same thing?

Reply:
Yes, I would recommend `hash.digest('hex')` here as the more direct API call. `hash.read()` is meant as a way of making a hash usable as a pipe. In order to even use the `read` command, for example, you are supposed to wait for the `readable` event.

Reply:
Updated!

Reply:
+ thanks for talking through the difference.